Semantic skeletons design (#1067)
unicode-org/message-format-wg: exploration/semantic-skeletons.md

# Semantic Skeletons Design

Status: **Proposed**

<details>
	<summary>Metadata</summary>
	<dl>
		<dt>Contributors</dt>
		<dd>@sffc</dd>
    <dd>@aphillips</dd>
		<dt>First proposed</dt>
		<dd>2024-04-06</dd>
		<dt>Pull Requests</dt>

> 		<dd>#000</dd>

Reviewer: ```suggestion
		<dd><a href="https://github.com/unicode-org/message-format-wg/pull/1067">#1067</a></dd>
```

	</dl>
</details>

## Objective

_What is this proposal trying to achieve?_

"Semantic skeletons" are a method introduced in CLDR 46 for programmatically selecting a datetime pattern for formatting. 
There is a fixed set of acceptable semantic skeletons.

Advantages of semantic skeletons over classical skeletons:

- A smaller set of acceptable options focused on producing outputs that are sensical
- Allows for a more efficient implementation
- Allows for a more clear, ergonomic API

> - More future-proof since CLDR recently added them (??)

Reviewer: This line of argumentation makes no sense. The new shiny thing being new and shiny does not mean that it'll last as long as tried and tested solutions.

Author: Removed the last argument. Clarified the earlier ones.

Advantages of semantic skeletons over other mechanisms:

> - Unlike "picture strings" (`MMM dd, yyyy`), skeletons do not require translation.

Reviewer: This is a straw-man argument, as picture strings have not at any point been even suggested for MF2. That should at least be noted somewhere.

Author: I can note that. Actually, picture strings were considered in the way-long ago. This text was added to acknowledge their existence (especially in MF1!) and document why we never consider them. There was a discussion of them in yesterday (2025-04-22)'s call which prompted these latest edits. 

I will add a note to point out the WG consensus not to support picture strings.

Author: Done

  Many date/time formatting regimes provide for "picture strings", which are patterns that
  exactly match the expected format of the output.
  In a picture string, separators, spaces, and other formatting are explicitly specified.
  This provides a lot of power to the devleoper or user experience designer, in terms of specifying formatting.
  However, this means that the translator has to modify the string to get localized output.
  For example, here are some picture strings with their output vs. the skeleton `MMMddyyyy`:
```
en-US
MMM dd, yyyy=Apr 21, 2025
dd MMM, yyyy=21 Apr, 2025
MM/dd/yyyy=04/21/2025
dd-MM-yyyy=21-04-2025
Apr 21, 2025
---
fr-FR
MMM dd, yyyy=avr. 21, 2025
dd MMM, yyyy=21 avr., 2025
MM/dd/yyyy=04/21/2025
dd-MM-yyyy=21-04-2025
21 avr. 2025
---
ja-JP
MMM dd, yyyy=4月 21, 2025
dd MMM, yyyy=21 4月, 2025
MM/dd/yyyy=04/21/2025
dd-MM-yyyy=21-04-2025
2025年4月21日

> ```

Reviewer: This should really be a table.

Author: done

Reviewer 2: I'm not sure why all the skeletons have 'yyyy' at the end even for Japanese. In Japanese, the year always comes at the start. Something like "4月 21, 2025" looks like a mess, and 04/21/2025 or 21-04-2025 will most probably be misunderstood one way or another. "2025年4月21日" or "2025/4/21" or something similar will all be easily understood.

Author: @duerst That's the point: they are *not* skeletons in the table. They are picture strings. Picture string produce wrong outcomes.

See [here](https://unicode-org.github.io/icu/userguide/format_parse/datetime/#date-field-symbol-table)
The system then uses the string to perform date/time pattern generation,
arranging the specified fields in the correct order,
selecting locale-appropriate separators,
and producing a "picture string" that can be consumed by date/time formatters
such as `java.text.SimpleDateFormat`.

## Use-Cases

_What use-cases do we see? Ideally, quote concrete examples._

As a developer, I want to format date, time, or date/time values to show specific fields
with specific appearance without having to learn a complex microsyntax
or modify my code or formatting directions for each locale.

As a translator, I want to understand what output a given date/time placeholder will produce 
in my language.

As a translator, I don't want to have to "translate" or modify a date/time placeholder to suit
my language's needs.
I should trust that the placeholder will produce appropriate results for my language.

## Requirements

_What properties does the solution have to manifest to enable the use-cases above?_

1. It should be possible to format common incremental time types

>    (e.g. milliseconds since epoch times)

Reviewer: This is an odd one to have as the very first requirement. I would challenge that it's _at all_ common to need to format epoch milliseconds in a localizable manner.

Author: "Incremental time types" are things like Javascript `Date`, Java `java.util.Date`, `time_t`, etc.

I would find it extremely odd if such types could not be formatted. I should change the parenthetical to be clearer. Formatting random `long` values is not what I meant.

Reviewer: Ah; this is meant to be about input, rather than output. The phrasing confused me.

Reviewer 3: I find this to be a requirement looking for a problem. _Of course_ if a language has a data type that represents a date, it should be able to be formatted with a MessageFormat date formatter.

Author: @sffc It might be possible to interpret this as only supporting Temporal values. It does no harm to spell out the need to format the various kinds of "classical" timekeeping.

2. It should be possible to format field-based time types
   (e.g. those that contain seperate values per field type in a date/time, such as a year-month)
3. It should be possible to format [floating time](https://www.w3.org/TR/timezone/#dfn-floating-time) values

>    (e.g. those that are not tied to a specific time zone)

Reviewer: Why do we care about the input value's type? Are these not implementation concerns?

Author: These are implementation concerns. But they generate requirements for date/time formatting options.

Reviewer: Really? Could you give an example, because I don't see how.

Author: One such problem is the need to manage an input value's relationship to the timeline. While temporal types trying to pack all of the information into strongly typed objects, classical time values are usually timestamps (seconds or millis of epoch time). Formatting them requires a time zone (or at least an offset). Which means some _external_ way of expressing the zone when it is different from the runtime's default time zone.
```
Delivery time local to you and not this web server will be: {$timestamp :datetime timezone=$usersTimeZone}
```

There is also a need to be able to "float" a value: remove it from the timeline so that field values stay constant regardless of time zone. This involves _removing_ the time zone from the value, the equivalent of `ZonedDateTime.toLocalDateTime()` in Java:
```
Your birthday is: {$birthday :datetime offset=none} (option name made up on the spot)
```

4. Date/time formatters should not permit users to format fields that don't exist in the value

>    (e.g. the "month" of a time, the "hour" of a date)

Reviewer: Again, isn't this an implementation concern? I don't see what this has to do with the shape of the formatting options.

Reviewer 3: This is the main benefit of the semantic skeleton design, and I think points 1-4 can all be merged into this one. The point is that not all date types are the same and not all formatters are the same, and the date type needs to be able to expose all fields that the formatter needs.

5. Date/time formatters should not permit users to format bad combinations of fields
   (e.g. `MMMMmm` (month-minute), `yyyyjm` (year-hour-minute), etc.)
6. Date/time formatters should permit users to control or influence the width of indvidual fields
   in a manner similar to classical skeletons

>    (e.g. `yMMd` vs. `yMMMd` vs. `yMMMMd` => 04/06/2025 vs. Apr 6, 2025 vs. April 6, 2025)

Reviewer: Influence, maybe; control, no. For example, not all locales have a sensible "short" form for months that makes sense when formatting as a full date, and enforcing its use can produce bad output.

Author: Control is probably too strong and influence is perhaps too weak. See what you think of the updated...

7. Developers, translators, and UI designers should only have to learn a single "microsyntax" or set of options for date formatting.
   Such a syntax or option set should be easy to understand only from the placeholder.

>    Such a syntax or option set should not require translators to alter the values in most or all locales.

Reviewer: Developers, translators, and UI designers do not work in a vacuum outside MF2, and they presumably are already familiar with some microsyntax or date formatting options. Given that, this requirement seems to be asking for the most common choice to be selected, and that seems surprising.

		<dt>First proposed</dt>
		<dd>2024-04-06</dd>
		<dt>Pull Requests</dt>
		<dd><a href="https://github.com/unicode-org/message-format-wg/pull/1067">#1067</a></dd>
	</dl>
</details>

## Objective

_What is this proposal trying to achieve?_

### Provide support for formatting date/time values using semantic skeletons

"Semantic skeletons" are a method introduced in CLDR 46 for programmatically selecting a datetime pattern for formatting. 
There is a fixed set of acceptable semantic skeletons.

Previously, ICU MessageFormat provided support for "classical skeletons",
using a microsyntax derived from familiar picture strings (see below)
combined with code in ICU (`DateTimePatternGenerator`) to produce the desired date/time value format.
`Intl.DateTimeFormat` uses "option bags" to provide a similar capability.
A classical skeleton allowed users to express the desired fields and field widths in a formatted date/time value.
The runtime uses locale data to determine minutiae such as 
field-order, 
separators, 
spacing, 
field-length, 
etc. to produce the desired output.

Advantages of semantic skeletons over classical skeletons:

- Provides all and only those combinations that make sense

>    - Allows for more efficient implementation, since there is no need to support "crazy" combinations like "month-hour"

Reviewer 4: ```suggestion
   - Allows for more efficient implementation, since there is no need to support unlikely combinations like "month-hour"
```

Avoid quoted adjectives in favor of a more precise description, if possible. Is it really true that _no one_ would want to use these combinations, or (my guess) is it more likely to be an error than not?

Author: done

- Allows for a more clear, ergonomic placeholder syntax, since the number of options can be limited
- Easier for user experience designers to specify, developers to implement, and translators to interpret

### Avoid 'picture strings'

The MFWG early on considered including support for "picture strings" in the formatting of date/time values.
There is a Working Group consensus **_not_** to support picture strings in Unicode MessageFormat, if possible.
Many date/time formatting regimes provide for "picture strings".
A "picture string" is a pattern using a microsyntax in which the user (developer, translator, UX designer)
exactly specifies the desired format of the date/time value.
In a picture string, separators, spaces, and other formatting are explicitly specified.

> This provides a lot of power to the devleoper or user experience designer, in terms of specifying formatting.

Reviewer 4: ```suggestion
This provides a lot of power to the developer or user experience designer, in terms of specifying formatting.
```

and producing a "picture string" that can be consumed by date/time formatters
such as `java.text.SimpleDateFormat`.

## Use-Cases

_What use-cases do we see? Ideally, quote concrete examples._

As a developer, I want to format date, time, or date/time values to show specific fields
with specific appearance without having to learn a complex microsyntax
or modify my code or formatting directions for each locale.

As a translator, I want to understand what output a given date/time placeholder will produce 
in my language.

As a translator, I don't want to have to "translate" or modify a date/time placeholder to suit
my language's needs.
I should trust that the placeholder will produce appropriate results for my language.

## Requirements

_What properties does the solution have to manifest to enable the use-cases above?_

1. It should be possible to format operands consisting of locally-relevant date/time types, including:
   - Temporal values such as `java.time` or JS `Temporal` values,
   - incremental time types ("timestamps")
     (e.g. milliseconds since epoch times such as `java.util.Date`, `time_t`, JS `Date`, etc.),
   - field-based time types

>      (e.g. those that contain seperate values per field type in a date/time, such as a year-month),

Reviewer 4: ```suggestion
     (e.g. those that contain separate values per field type in a date/time, such as a year-month),
```

   while relying on locale data to prevent undesirable results.
   For example:
   | Classical Skeleton | `en-US` Output | 
   |---|---|
   | `yMd` | 04/06/2025 |
   | `yMMMd` | Apr 6, 2025 |
   | `yMMMMd` | April 6, 2025 |
1. Developers, translators, and UI designers should not have to learn
   multiple new microsyntaxes or multiple different sets of options for date/time value formatting.
1. Any microsyntax or option set specified should be easy to understand only from the placeholder.
1. Any microsyntax or option set specified should not _require_ translators to alter the values in most or all locales.

## Constraints

_What prior decisions and existing conditions limit the possible design?_

## Proposed Design

_Describe the proposed solution. Consider syntax, formatting, errors, registry, tooling, interchange._

## Alternatives Considered

_What other solutions are available?_
_How do they compare against the requirements?_
_What other properties they have?_

### Design: Use Option Naming

In this section, we use a scheme similar to `FieldSetBuilder` linked earlier.

#### DateTime fields

Options:

```
{$date :datetime dateFields="YMD"}  
{$date :datetime date="YMD"}  

> {$date :datetime fields="YMD"}

Reviewer 4: Are `dateFields`, `date` and `fields` just different possible names for the same option? Or do they mean different things? (Same question about `timePrecision` vs. `time` below.)

Reviewer 3: Different possible names for the same option.

> ```

Reviewer 4: ```suggestion
```

The names `dateFields`, `date`, and `fields` are candidate names for the option that specifies the semantic skeleton string to be used for formatting the date/time value.
```

#### TimePrecision

> 

Reviewer 4: I think this needs a little more explanation (it should be possible to follow this doc without reading the other linked-to docs).

Author: This part of the spec is basically empty and in need of work. I started to noodle around with it today. Each one of the options will need quasi-complete descriptions in order to see how they'll work and the relative usability of each.

etc. to produce the desired output.

Advantages of semantic skeletons over classical skeletons:

- Provides all and only those combinations that make sense
   - Allows for more efficient implementation, since there is no need to support combinations like "month-hour"
     that are not useful to users
- Allows for a more clear, ergonomic placeholder syntax, since the number of options can be limited
- Easier for user experience designers to specify, developers to implement, and translators to interpret

#### Avoid 'picture strings'

The MFWG early on considered including support for "picture strings" in the formatting of date/time values.
There is a Working Group consensus **_not_** to support picture strings in Unicode MessageFormat, if possible.
Many date/time formatting regimes provide for "picture strings".
A "picture string" is a pattern using a microsyntax in which the user (developer, translator, UX designer)
exactly specifies the desired format of the date/time value.
In a picture string, separators, spaces, and other formatting are explicitly specified.
This provides a lot of power to the developer or user experience designer, in terms of specifying formatting.

> For example: `MMM dd, yyyy` or `yyyy-dd-MM'T'HH:mm:ss`

Author: Fair enough. I moved the WG bits on top to make them visible, but this works.

in my language.

As a translator, I don't want to have to "translate" or modify a date/time placeholder to suit
my language's needs.
I should trust that the placeholder will produce appropriate results for my language.

## Requirements

_What properties does the solution have to manifest to enable the use-cases above?_

Users want the most intuitive formats and outcomes to be the defaults.
They should not have to coerce or convert normal date/time types in order to do simple operations.

1. It should be possible to format operands consisting of locally-relevant date/time types, including:
   - Temporal values such as `java.time` or JS `Temporal` values,
   - incremental time types ("timestamps")
     (e.g. milliseconds since epoch times such as `java.util.Date`, `time_t`, JS `Date`, etc.),
   - field-based time types
     (e.g. those that contain separate values per field type in a date/time, such as a year-month),
   - [floating time](https://www.w3.org/TR/timezone/#dfn-floating-time) values
     (e.g. those that are not tied to a specific time zone, variously called local/plain/civil times),
   -  or other local exotica (Java `Calendar`, C `tm` struct, etc.)

> 1. Date/time formatters should not permit users to format fields that don't exist in the value
>    (e.g. the "month" of a time, the "hour" of a date)

Reviewer: This presumably should not be understood to mean that e.g. epoch milliseconds ought not be formattable as year-month-date, given that the value only contains milliseconds, yes?

```suggestion
1. Date/time formatter options should be able to impose restrictions on acceptable values,
   such as formatting a month requiring a date, and formatting an hour requiring a time.
```

Author: They're not really "formatter options" though, are they? `{$timeOnlyValue :date}` == Bad Operand?

```suggestion
1. Date/time formatting functions and their options should be able to impose restrictions on acceptable values,
   such as formatting a month requiring a value that includes the month 
   or formatting an hour requiring requiring a time.
1. Date/time formatting functions and their options should be able to emit errors 
   when trying to format fields that cannot be computed from the operand value
   such as the "month" of a time-only value (`{|03:45Z| :date format=YMD}`)
   or the "hour" of a date-only value (`{|2025-04-23| :time format=HM}`).
```

Reviewer: Your suggested second requirement smells like scope creep to me. I understand this design doc to be about function options, not about enabling or restricting a function handler from emitting errors.

We don't actually impose any restriction on when a function handler can emit an error (just on what errors may be emitted when), we already allow an implementation with an options-basket or even classical-skeleton option style to reject option combinations that could not be expressed as a semantic skeleton.

This design doc should be focused on the ergonomics and readability of the different options styles, and not on implementation details.

1. Date/time formatters should not permit users to format bad combinations of fields
   (e.g. `MMMMmm` (month-minute), `yyyyjm` (year-hour-minute), etc.)
1. Date/time formatters should permit users to specify the desired width of indvidual fields
   in a manner similar to classical skeletons,

>    while relying on locale data to prevent undesirable results.

Reviewer 3: I want to be more clear about what is "required" versus "wanted".

I think "required" is that users should specify an overall length, and "wanted" to hint at the width of an individual field independently of the overall length.

   For example:
   | Classical Skeleton | `en-US` Output | 
   |---|---|
   | `yMd` | 04/06/2025 |
   | `yMMMd` | Apr 6, 2025 |
   | `yMMMMd` | April 6, 2025 |
1. Developers, translators, and UI designers should not have to learn
   multiple new microsyntaxes or multiple different sets of options for date/time value formatting.

> 1. Any microsyntax or option set specified should be easy to understand only from the placeholder.

Reviewer: ```suggestion
1. Any microsyntax or option set specified should be easy to understand only from the expression.
```

_What other properties they have?_

### Design: Use Option Naming

In this section, we use a scheme similar to `FieldSetBuilder` linked earlier.

#### DateTime fields

Options:

```
{$date :datetime dateFields=YMD}  
{$date :datetime date=YMD}  
{$date :datetime fields=YMD}
```

The names `dateFields`, `date`, and `fields` are candidate names for the option that specifies the semantic skeleton string to be used for formatting the date/time value.
#### TimePrecision

Options:
```
{$date :datetime timePrecision=minute}  
{$date :datetime time=minute}
```
(TODO: Add others)

### Design: Use Separate Functions

Some choices:

#### A single `:datetime` function

_Pros_
- Everything is in one place

_Cons_
- More combinations of options that can form invalid skeletons
- Can require verbose placeholders

#### Use separate `:date`, `:time` and `:datetime` functions

_Pros_
- More tailored to specific user requirements
- `:date` and `:time` are somewhat self-documenting about the message author's formatting intention

_Cons_
- Not fully type-safe.

#### Use separate semantic functions

`:date`, `:time`, `:datetime`, `:zoneddatetime`, *maybe* `:zoneddate`, `:zonedtime`, `:timezone`

> Problem: Most users are likely to prefer date/time/datetime to zoneddate/zonedtime/zoneddatetime

Reviewer 3: Great, so if naming is the only problem here, let's bikeshed the names and make another set of names. Let's try not to deviate _too much_ from java.time and Temporal. :)

as the default formatting functions.
Most "classical" time types are timestamps.
User intentions might not be met by these names.

Problem: Different platforms cannot agree on what to call a Floating Time Value: HTML and Java use `LocalXXX`,
JavaScript has adopted `PlainXXX`,

> some others use different terms, such as `CivilXXX`.

Reviewer: I don't understand the relevance of this. Isn't this purely an implementation concern, which has no visibility in the MF2 syntax? As I understand it, the idea is for the function names to describe the formatted output, not the input.

Author: A reasonable approach would be to name the functions consistent with usage. The point here is that there is not agreement on what to call these.

Reviewer: I still don't understand how naming concerns about an input value type would transmit to the syntax. For example, I would expect an expression for a time with hours and minutes to look something like
```
{$t :datetime hour=numeric minute=numeric}
```
or maybe
```
{$t :time fields=hm}
```
irrespective of whether `$t` held a floating time value, or one with a timezone attached, or any other representation of the value to be formatted.

Or do you think we ought to have different functions for each input datetime type supported by an implementation?

Reviewer 3: By "type safe" I mean that there should be well-defined semantics around when implementations should emit a Bad Operand error. For example, if you pass an implementation-specific "time-only" type to `:date`, you should get a Bad Operand error.

_Pros_
- Helps users get the right results
- Could be less verbose
- Better at documenting the message author's intention

_Cons_

> - _Lots_ of functions

Reviewer 3: But only 1 more than the previous option. The only one this adds is `:zoneddatetime`.

   -  or other local exotica (Java `Calendar`, C `tm` struct, etc.)
1. Date/time formatter options should be able to impose restrictions on acceptable values,
   such as formatting a month requiring a date, and formatting an hour requiring a time.
1. Date/time formatters should not permit users to format bad combinations of fields
   (e.g. `MMMMmm` (month-minute), `yyyyjm` (year-hour-minute), etc.)
1. Date/time formatters should permit users to specify the desired width of indvidual fields
   in a manner similar to classical skeletons,
   while relying on locale data to prevent undesirable results.
   For example:
   | Classical Skeleton | `en-US` Output | 
   |---|---|
   | `yMd` | 04/06/2025 |
   | `yMMMd` | Apr 6, 2025 |
   | `yMMMMd` | April 6, 2025 |
1. Developers, translators, and UI designers should not have to learn
   multiple new microsyntaxes or multiple different sets of options for date/time value formatting.
1. Any microsyntax or option set specified should be easy to understand only from the expression.
1. Any microsyntax or option set specified should not _require_ translators to alter the values in most or all locales.

> 1. User's should be able to specify the time zone or local time offset used in formatting
>    an incremental or timestamp value
>    because many date/time formatting processes are
>    running in a different environment/offset/zone from where the value will be seen
>    or are formatting a value for a specific offset or zone.

Reviewer: I agree with being able to specify the timezone, either while constructing the formattable value, or possibly within the _options_, but I don't see the use case for explicit offset support. Could we drop that?
```suggestion
1. Users should be able to specify the timezone used in formatting
   an incremental or timestamp value
   because many date/time formatting processes are
   running in a different environment or timezone from where the value will be seen
   or are formatting a value for a specific timezone.
```

Author: As I said, it's not optimal to use offset. But offsets are common. If we made an optional option for them, that's less interoperable. Documentation should definitely discourage offsets vs. true zones.  But their prevalence is high enough that I think support might end up being necessary. Note that temporal has offset types.

An alternative would be using offset zones?  (I know... ick...)

We should start to put option sets together to compare, much as you've been doing in some comments. I can't do that from my phone.

Reviewer: Offset zones aren't a full solution, because they're only available with integer hour values.

Regarding "necessary", note that AFAIK no extant message formatting syntax allows for setting the timezone or offset within the message syntax, and that all message formatting happens in the context of some programming language environment. In other words, we are here introducing a novel capability, and none of this is strictly speaking necessary -- it would be perfectly valid for us to rely solely on the timezone or offset to be set in the code that sets the datetime variable value that the formatter is called with.

Also, we don't need to introduce all capabilities at once. If we leave out something e.g. from the 48 release, it can be added in later. The reverse is not true.

Author: > offset zones aren't a full solution, because they're only available with integer hour values.

Offset zones are not restricted to integer hour values. The `Etc/GMT` zones built into systems are often integer or half-integer hour zones or such, but this is not a universal restriction. I'm not really talking about using constructed zones, in any case.

> note that AFAIK no extant message formatting syntax allows for setting the timezone or offset within the message syntax

Most message formatters require the caller to manage the time zone in code and there exist API interfaces. You might recall that I suggested that time zone be part of the formatting context. Lots of the problems that developers have with floating time values stem from the fact that there is poor control at the message level and developers have to cobble together solutions.

> Also, we don't need to introduce all capabilities at once. If we leave out something e.g. from the 48 release, it can be added in later. The reverse is not true.

My "necessary" sentence was fairly _suppositional_. I am leery of incomplete solutions because they lead to bad coding or formatting patterns. I agree that we can omit things that are not well-baked or for which the syntax and features are not yet well-understood (or are _suppositional_, if you prefer 🤣). I'd rather not deprecate things later/ever, if we can avoid it.

However, my point was that we have a good bit of experience formatting date/time values. In this document we're assembling requirements and trying to define the set of functions and options. It's good to at least look at the use cases--to put all of the utensils on the table, not just the fish forks and teaspoons--before deciding a design.

> To clarify, I'm not objecting at all to supporting the formatting of input values that include an offset, but explicitly setting an offset within an MF2 function call seems like a capability that ought to be left to implementations that choose to support it, rather than something to require of all implementations.

In any case, your original suggestion is about requirement text. I can see how we might make that clearer without the requirement requiring offsets be supported. Perhaps:

```suggestion
1. Message authors and developers should be able to manage 
   the formatting of date/time values such that
   processes running in an environment with a different time zone or local time offset
   from that needed to express or format the value
   can get the necessary results.
```

Reviewer: That suggestion looks good; you'll probably need to re-apply it separately as the typo fix has made this thread "outdated".

  - plus the common combo `:time` and `:date`
- Date Field Sets
  - `:day-weekday`, `:month-day`, `:month-day-weekday`, `:year-month-day`, `:year-month-day-weekday`
- Composite field sets
  - `:date-zone`, `:date-time-zone`, `:time-zone` (note: `time-zone` is two fields, not "timezone")

Examples:
```
Your package arrived on {$d :date} at {$d :time}.
    Your package arrived on Apr 27, 2025 at 10:50 AM.
Your package arrived on {$d :month-day-weekday} at {$d :time-zone}.
    Your package arrived on Sunday, April 27th at 10:50 AM PDT.
Your package arrived on {$d :month-day-weekday month=medium weekday=full} at {$d :time-zone display=full zone=short}.
    Your package arrive on Sunday, Apr 27th at 10:50 PDT.
```

_Pros_
- Options focused on field widths
- Self-documenting: obvious what fields are shown
- Reserves options for field width

_Cons_
- _Nineteen_ (or so) functions
- Names appear generative, but aren't actually
- Options might be baroque. Some option values might not be compatible with one another.
  - Expression-wide options (short/medium/long/full) and field-level options might both be needed together

> 

Reviewer: For example:
```js
new Intl.DateTimeFormat('fi', { month: 'long' }).format()
// 'toukokuu'

new Intl.DateTimeFormat('fi', { day: 'numeric', month: 'long' }).format()
// '2. toukokuuta'
```

```suggestion
- Leads to localization failures similar to picture strings,
  as combinations of standalone fields are formatted incorrectly in inflected languages.
```

Author: I don't understand this addition. I mean, I understand what you mean, but not why this is specific to this solution. Any skeleton solution should produce the correct output as a result of running the date/time pattern generator on the selected combination of fields and field widths. FWIW, all solutions can produce incorrect results if the information about the desired format is incomplete.

Reviewer: My concern is about the affordances we provide. Something like
```
{$d :month}/{$d :day}
```
is much likelier to get actual use than
```
{$d :datetime month=numeric}/{$d :datetime day=numeric}
```
or
```
{$d :date fields=month}/{$d :date fields=day}
```
because it's much shorter, and doesn't immediately communicate to the reader how hacky it is. Meanwhile, the two latter choices hint much more that this isn't good prcatice.

Author: Okay, that's a good callout. I think we might make standalone fields less convenient across the board for similar reasons. We enable them, of course, but when you need them they should require more work?

Reviewer: I wouldn't say that we "make them less convenient", but more like we don't specifically make them excessively convenient, but this is semantics and I think we agree.

Reviewer: ```suggestion
- Providing easy-to-use standalone fields may lead to localization failures similar to picture strings,
  as their combinations are formatted incorrectly in inflected languages.
```

Author: I'm going to modify this slightly to include the examples for posterity or future use in our own documentation.

```suggestion
- Providing easy-to-use standalone field functions 
  could lead to localization failures similar to picture strings,
  as naive users might use combinations of standalone functions to concatenate date/time values
  that are inappropriate in various other locales and require translator intervention.
> For example, this message hard codes separators, field order, and
> uses standalone representation of the fields:
> ```
> Today is {$d :month} {$d :day}, {$d :year} and it is {$d :hour}:{$d :minute} o'clock.
>```
```

Author: Also, note that this is an objection to **any** standalone formats and might be better located in requirements/use cases with a short con for this item (and other designs that bump up against it)

Reviewer: Modified suggestion looks good; I edited it to indent the `>` block.

combined with code in ICU (`DateTimePatternGenerator`) to produce the desired date/time value format.

`Intl.DateTimeFormat` provided options to provide a similar capability.
These weren't "skeletons" from the point of view that they didn't use a dedicated microsyntax
similar to 'picture strings',
but the effect was the same:

> users specified which fields they wanted with which display options (such as width).

Reviewer: `Intl.DateTimeFormat` is still very much a thing.

```suggestion
`Intl.DateTimeFormat` provides options to provide a similar capability.
These aren't "skeletons" from the point of view that they don't use a dedicated microsyntax
similar to 'picture strings',
but the effect is the same:
users specify which fields they want with which display options (such as width).
```

and producing a "picture string" that can be consumed by date/time formatters
such as `java.text.SimpleDateFormat`.

### FAQ

This section considers some potential arguments against the design
or captures frequently asked questions about it.

> **What if semantic skeletons doesn't support the format I want?
> Unlike picture strings or classical skeletons, semantic skeletons do not allow unrestricted
> composition of date/time formats.**
> 

Reviewer: ```suggestion
**What if semantic skeletons don't support the format I want?**

Unlike picture strings or classical skeletons,
semantic skeletons do not allow unrestricted composition of date/time formats.

```

However, the designers of semantic skeletons have considered the breadth of use cases.
If a skeleton is not available, there is probably a good reason for avoiding it.

**My UX designer wants to specify different separators/presentational details.
I could do that with picture strings, but not with any kind of skeleton. Help!**

Specialized formats might be constructed using individual fields or by using formatToParts.
In general, such specialized designs rapidly become examples of poor internationalization,
since examples of such adjustments do not consider the breadth of date/time representation.

**Semantic skeletons are too new. 
Implementation experience is limited.
Shouldn't we wait to adopt them?**

Unicode MessageFormat has a one-time opportunity to avoid "deprecated at birth" date/time formatting
and to provide a robust, internally-consistent mechanism that guides users away from
common date/time formatting pitfalls.

This is already ample experience with classical skeletons.
The difference with semantic skeletons is that it filters out "mistakes"
such as "11 PM April" (`jjMMMM` or `HHaMMMM`) or "2 2025" (`dyyyy`).

**What about specialized formats, such as ISO8601?**

> These should be provided via other means that requiring a specialized pattern 
> and the (optional) `@locale` attribute. Do you really want to support this,
> given that it can then be used for other things:
> ```
> {$now :datetime pattern=|yyyy-MM-dd'T'HH:mm:ss.sssz| @locale=und timezone=UTC}
> ```

Reviewer: ````suggestion
These should be provided via other means than requiring a specialized pattern 
and the (optional) `u:locale` option.
Do you really want to support this,
given that it can then be used for other things:
```
{$now :datetime pattern=|yyyy-MM-dd'T'HH:mm:ss.sssz| u:locale=und timezone=UTC}
```
````
